### PR TITLE
Add test for auth

### PR DESCRIPTION
**Pull request by ghazalrafiei**

### CODE REVIEW

File: auth/auth_test.go [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+		{ConfigPath: cp[6], Pass: false},
+		{ConfigPath: cp[7], Pass: false},
+	}
+	for tc, _ := range testCases {
+		for tc, _ := range testCases {
+			config, err1 := config.NewConfig(testCases[tc].ConfigPath)
+			GetResult(testCases[tc].Pass, err1)
+			if err1 != nil {
+				PrintResult(t, testCases[tc].Pass, err1)
+			} else if _, err2 := auth.NewAuth(config); err2 != nil {
```
Comment (amiraminbeidokhti) [inline auth/auth_test.go:63]: Undefined auth.
You are in the same package, so you have to drop the package name or you cannot build this package.

File: auth/auth_test.go [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+func TestNewAuth(t *testing.T) {
+	testCases := []TestInput{
+		{ConfigPath: cp[0], Pass: true},
+		{ConfigPath: cp[1], Pass: true},
+		{ConfigPath: cp[2], Pass: true},
+		{ConfigPath: cp[5], Pass: false},
+		{ConfigPath: cp[6], Pass: false},
+		{ConfigPath: cp[7], Pass: false},
+	}
+	for tc, _ := range testCases {
```
Comment (amiraminbeidokhti) [inline auth/auth_test.go:57]: I cannot understand this part.
Why did you use nested loop? and for both of them declared tc variable.
A better approach is:
for _, test := range testCases {
    statements
}

File: auth/auth_test.go [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+	tip    TestInput
+	id     string
+	method object.Method
+	key    object.Key
+	Pass   bool
+}
+
+func TestHash(t *testing.T) {
+	testCases := []TestInput1{
+		{tip: TestInput{ConfigPath: cp[0]}, id: "agent-123", method: object.MethodGet, key: object.Key{Type: "world", Name: "judge", Namespace: "default"}, Pass: true},
```
Comment (amiraminbeidokhti) [inline auth/auth_test.go:170]: It's better to use the variables you declared at the end of this file - methodCases

File: auth/example/roles2.yaml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+          type: "world"
+        methods:
+          - "get"
+          - "set"
+          - "find"
+          - "watch"
+          - "delete"
+  - name: "logger"
+    rules:
+      - key:
```
Comment (amiraminbeidokhti) [inline auth/example/roles2.yaml:35]: Didn't find expected key

File: auth/auth_test.go (hunk 1 above)
Comment (ghazalrafiei) [reply auth/auth_test.go:63]: I had uploaded an older version of this file! It's fixed now.

File: auth/example/roles2.yaml (hunk 4 above)
Comment (ghazalrafiei) [reply auth/example/roles2.yaml:35]: I haven't bring this case in my test cases. What exactly do you mean?
I also didn't altered test files. I just copied them from files config and example.

File: auth/auth_test.go (hunk 2 above)
Comment (ghazalrafiei) [reply auth/auth_test.go:57]: Fixed

File: auth/auth_test.go [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+	object.Key{Type: "nonaction", Namespace: "default"},
+	object.Key{},
+	object.Key{Type: "world"},
+
+	object.Key{Type: "nonaction", Name: "wrongName", Namespace: "default"},
+	object.Key{Type: "nonaction"},
+	object.Key{Name: "wrongName", Namespace: "default"},
+}
+var objectCases = []object.Object{
+	//object.Object{},     expected to fail but the program exits (?)
```
Comment (amiraminbeidokhti) [inline auth/auth_test.go:219]: that would be better if you delete this line

Comment (ghazalrafiei) [reply auth/auth_test.go:219]: Shouldn't the program print a log on this case and continue? It just releases an error and terminates.

File: auth/auth_test.go [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+					PrintResult(t, test.Pass, nil)
+				}
+			}
+		}
+	}
+}
+
+var objectKeys = []object.Key{
+	//config0
+	object.Key{Type: "world", Name: "wrongName"},
```
Comment (alidadar7676) [inline auth/auth_test.go:195]: Remove redundant type definitions. I mean `object.Key` can be removed from these lines.

File: auth/auth_test.go [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+}
+var objectCases = []object.Object{
+	//object.Object{},     expected to fail but the program exits (?)
+	object.Object{Key: &objectKeys[0], Value: "Value"},
+	object.Object{Key: &objectKeys[1], Value: "testing"},
+	object.Object{Key: &objectKeys[2], Value: ""},
+	object.Object{Key: &objectKeys[3], Value: "Value"},
+	object.Object{Key: &objectKeys[4], Value: ""},
+	object.Object{Key: &objectKeys[5], Value: "abc"},
+	object.Object{Key: &objectKeys[6], Value: "abc"},
```
Comment (alidadar7676) [inline auth/auth_test.go:226]: Redundant type definition

File: auth/auth_test.go (hunk 5 above)
Comment (alidadar7676) [reply auth/auth_test.go:219]: It is a bug in the `auth` package. We should check the value of `obj.Key` and then use it

File: auth/auth_test.go [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+		{ConfigPath: cp[6], id: "logger-123", Pass: false},
+		{ConfigPath: cp[7], id: "", Pass: false}, //idk
+	}
+	for tc, _ := range testCases {
+		config, err1 := config.NewConfig(testCases[tc].ConfigPath)
+		if err1 != nil {
+			PrintResult(t, testCases[tc].Pass, err1)
+		} else {
+			au, err2 := NewAuth(config)
+			if err2 != nil {
```
[the remark this answers is not in the document]
Comment (ghazalrafiei) [reply auth/auth_test.go:142]: I thought we can not define a variable two times.

**Pull request merged**